File: Sources/BraintreeApplePay/BTConfiguration+ApplePay.m
#if __has_include(<Braintree/BraintreeApplePay.h>)
#import <Braintree/BTConfiguration+ApplePay.h>
#else
#import <BraintreeApplePay/BTConfiguration+ApplePay.h>
#endif

// Swift Module Imports
#if __has_include(<Braintree/Braintree-Swift.h>) // Cocoapods-generated Swift Header
#import <Braintree/Braintree-Swift.h>

#elif SWIFT_PACKAGE                              // SPM
/* Use @import for SPM support
 * See https://forums.swift.org/t/using-a-swift-package-in-a-mixed-swift-and-objective-c-project/27348
 */
@import BraintreeCoreSwift;

#elif __has_include("Braintree-Swift.h")         // CocoaPods for ReactNative
/* Use quoted style when importing Swift headers for ReactNative support
 * See https://github.com/braintree/braintree_ios/issues/671
 */
#import "Braintree-Swift.h"

#else // Carthage or Local Builds
#import <BraintreeCoreSwift/BraintreeCoreSwift-Swift.h>
#endif

@implementation BTConfiguration (ApplePay)

- (BOOL)isApplePayEnabled {
    BTJSON *applePayConfiguration = self.json[@"applePay"];
    return [applePayConfiguration[@"status"] isString] && ![[applePayConfiguration[@"status"] asString] isEqualToString:@"off"];
}

- (BOOL)canMakeApplePayPayments {
    return [PKPaymentAuthorizationViewController canMakePaymentsUsingNetworks:self.applePaySupportedNetworks];
}

- (NSString *)applePayCountryCode {
    return [self.json[@"applePay"][@"countryCode"] asString];
}

- (NSString *)applePayCurrencyCode {
    return [self.json[@"applePay"][@"currencyCode"] asString];
}

- (NSString *)applePayMerchantIdentifier {
    return [self.json[@"applePay"][@"merchantIdentifier"] asString];
}

- (NSArray<PKPaymentNetwork> *)applePaySupportedNetworks {
    NSArray <NSString *> *gatewaySupportedNetworks = [self.json[@"applePay"][@"supportedNetworks"] asStringArray];

    NSMutableArray <PKPaymentNetwork> *supportedNetworks = [NSMutableArray new];
    
    for (NSString *gatewaySupportedNetwork in gatewaySupportedNetworks) {
        if ([gatewaySupportedNetwork localizedCaseInsensitiveCompare:@"visa"] == NSOrderedSame) {
            [supportedNetworks addObject:PKPaymentNetworkVisa];
        } else if ([gatewaySupportedNetwork localizedCaseInsensitiveCompare:@"mastercard"] == NSOrderedSame) {
            [supportedNetworks addObject:PKPaymentNetworkMasterCard];
        } else if ([gatewaySupportedNetwork localizedCaseInsensitiveCompare:@"amex"] == NSOrderedSame) {
            [supportedNetworks addObject:PKPaymentNetworkAmex];
        } else if ([gatewaySupportedNetwork localizedCaseInsensitiveCompare:@"discover"] == NSOrderedSame) {
            [supportedNetworks addObject:PKPaymentNetworkDiscover];
        } else if ([gatewaySupportedNetwork localizedCaseInsensitiveCompare:@"maestro"] == NSOrderedSame) {
            [supportedNetworks addObject:PKPaymentNetworkMaestro];
        } else if ([gatewaySupportedNetwork localizedCaseInsensitiveCompare:@"elo"] == NSOrderedSame) {
            [supportedNetworks addObject:PKPaymentNetworkElo];
        }
    }
    return [supportedNetworks copy];
}

@end

File: Sources/BraintreeApplePay/BTConfiguration+ApplePay.swift
import Foundation
import PassKit

#if canImport(BraintreeCore)
import BraintreeCore
#endif

extension BTConfiguration {

    /// :nodoc: This method is exposed for internal Braintree use only. Do not use. It is not covered by Semantic Versioning and may change or be removed at any time.
    /// Indicates whether Apple Pay is enabled for your merchant account.
    @_documentation(visibility: private)
    @objc public var isApplePayEnabled: Bool {
        guard let applePayConfiguration: BTJSON = json?["applePay"] else { return false }
        return applePayConfiguration["status"].isString && applePayConfiguration["status"].asString() != "off"
    }

    /// Indicates if the Apple Pay merchant enabled payment networks are supported on this device.
    var canMakeApplePayPayments: Bool {
        guard let applePaySupportedNetworks else { return false }
        return PKPaymentAuthorizationViewController.canMakePayments(usingNetworks: applePaySupportedNetworks)
    }

    /// The country code for your Braintree merchant account.
    var applePayCountryCode: String? {
        json?["applePay"]["countryCode"].asString()
    }

    /// The Apple Pay currency code supported by your Braintree merchant account.
    var applePayCurrencyCode: String? {
        json?["applePay"]["currencyCode"].asString()
    }

    /// The Apple Pay merchant identifier associated with your Braintree merchant account.
    var applePayMerchantIdentifier: String? {
        json?["applePay"]["merchantIdentifier"].asString()
    }

    /// The Apple Pay payment networks supported by your Braintree merchant account.
    var applePaySupportedNetworks: [PKPaymentNetwork]? {
        let gatewaySupportedNetworks: [String]? = json?["applePay"]["supportedNetworks"].asStringArray()
        var supportedNetworks: [PKPaymentNetwork] = []

        gatewaySupportedNetworks?.forEach { gatewaySupportedNetwork in
            if gatewaySupportedNetwork.localizedCaseInsensitiveCompare("visa") == .orderedSame {
                supportedNetworks.append(.visa)
            } else if gatewaySupportedNetwork.localizedCaseInsensitiveCompare("mastercard") == .orderedSame {
                supportedNetworks.append(.masterCard)
            } else if gatewaySupportedNetwork.localizedCaseInsensitiveCompare("amex") == .orderedSame {
                supportedNetworks.append(.amex)
            } else if gatewaySupportedNetwork.localizedCaseInsensitiveCompare("discover") == .orderedSame {
                supportedNetworks.append(.discover)
            } else if gatewaySupportedNetwork.localizedCaseInsensitiveCompare("maestro") == .orderedSame {
                supportedNetworks.append(.maestro)
            } else if gatewaySupportedNetwork.localizedCaseInsensitiveCompare("elo") == .orderedSame {
                supportedNetworks.append(.elo)
            }
        }

        return supportedNetworks
    }
}
